src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1578,6 +1584,80 @@ private static bool HasAnonymousFunctionConversion(BoundExpression source, TypeS
 
             return IsAnonymousFunctionCompatibleWithType((UnboundLambda)source, destination) == LambdaConversionResult.Success;
         }
+
+        internal enum ConstructibleCollectionTypeKind
+        {
+            None = 0,
+            Array,
+            Span,
+            ReadOnlySpan,
+            CollectionInitializer,
+        }
+
+        internal static ConstructibleCollectionTypeKind GetConstructibleCollectionType(CSharpCompilation compilation, TypeSymbol destination, out TypeSymbol? elementType)
+        {
+            Debug.Assert(compilation is { });
+
+            if (destination is ArrayTypeSymbol arrayType)
+            {
+                if (arrayType.IsSZArray)
+                {
+                    elementType = arrayType.ElementType;
+                    return ConstructibleCollectionTypeKind.Array;
+                }
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_Span_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.Span;
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_ReadOnlySpan_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.ReadOnlySpan;
+            }
+            else if (supportsCollectionInitializer(compilation, destination))
+            {
+                elementType = null;
+                return ConstructibleCollectionTypeKind.CollectionInitializer;
+            }
+
+            elementType = null;
+            return ConstructibleCollectionTypeKind.None;
+
+            static bool isSpanType(CSharpCompilation compilation, TypeSymbol targetType, WellKnownType spanType, [NotNullWhen(true)] out TypeSymbol? elementType)
+            {
+                if (targetType is NamedTypeSymbol { Arity: 1 } namedType
+                    && TypeSymbol.Equals(namedType.OriginalDefinition, compilation.GetWellKnownType(spanType), TypeCompareKind.AllIgnoreOptions))
+                {
+                    elementType = namedType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+                    return true;
+                }
+                elementType = null;
+                return false;
+            }
+
+            static bool supportsCollectionInitializer(CSharpCompilation compilation, TypeSymbol targetType)
```

Comment:
This is missing the `Add` checks on whether a type supports collection initializers. Perhaps rename this to just `typeImplementsIEnumerable`?